experiments/MoarCSets/SInstances.jl

```diff
@@ -0,0 +1,406 @@
+module SInstances
+export Schema, SchemaExpr, DataExpr, AttrExpr, FreeSchema,
+  CatDesc, CatDescType, AttrDesc, AttrDescType, AbstractSInstance, SInstance,
+  nparts, add_part!, add_parts!, _add_parts!, subpart, incident, set_subpart!, set_subparts!,
+  TheoryGraph, AbstractGraph, Graph,
+  TheoryDecGraph, AbstractDecGraph, DecGraph
+
+using Catlab.GAT, Catlab.Syntax, Catlab.Present, Catlab.Theories
+using StructArrays
+using MLStyle
+
+@theory Category(Ob,Hom) => Schema(Ob,Hom,Data,Attr) begin
+  Data::TYPE
+  Attr(dom::Ob,codom::Data)::TYPE
+
+  compose(f::Hom(A,B), g::Attr(B,X))::Attr(A,X) ⊣ (A::Ob, B::Ob, X::Data)
+  (compose(f, compose(g, a)) == compose(compose(f, g), a)
+    ⊣ (A::Ob, B::Ob, C::Ob, X::Data, f::Hom(A,B), g::Hom(B,C), a::Attr(C, X)))
+  compose(id(A), a) == a ⊣ (A::Ob, X::Ob, a::Attr(A,X))
+end
+
+abstract type SchemaExpr{T} <: GATExpr{T} end
+abstract type DataExpr{T} <: SchemaExpr{T} end
+abstract type AttrExpr{T} <: SchemaExpr{T} end
+
+@syntax FreeSchema(ObExpr,HomExpr,DataExpr,AttrExpr) Schema begin
+  # should have a normal representation for precompose of a morphism + a generator attribute
+end
+
+struct CatDesc{Ob,Hom,Dom,Codom}
+  function CatDesc{Ob,Hom,Dom,Codom}() where
+    {Ob,Hom,Dom,Codom}
+    new{Ob,Hom,Dom,Codom}()
+  end
+  function CatDesc(pres::Presentation{Schema})
+    obs, homs = generators(pres, :Ob), generators(pres,:Hom)
+    ob_syms, hom_syms = nameof.(obs), nameof.(homs)
+    ob_num = ob -> findfirst(ob_syms .== ob)::Int
+    new{Tuple(ob_syms), Tuple(hom_syms),
+        Tuple(@. ob_num(nameof(dom(homs)))), Tuple(@. ob_num(nameof(codom(homs))))}()
+  end
+end
+
+CatDescType(pres::Presentation{Schema}) = typeof(CatDesc(pres))
+
+function Base.getproperty(AD::Type{T},i::Symbol) where
```

Comment:
I don't know if this technically counts as ["type piracy"](https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-type-piracy-1) because we "own" the type `CatDesc`, nevertheless this smells funny, especially having to call `invoke` to simulate a "super" call in a language where that pattern is more standard. I think we should consider a different approach, even if means having to make a few more keystrokes, say `typeparam(T, :ob)`.

**Update**: Instead of calling `invoke`, we can fall back to `getfield`, see this [discourse thread](https://discourse.julialang.org/t/whats-the-difference-between-fields-and-properties/12495/2). That should be an adequate fix, so we can keep this nice syntactic sugar.

Reply:
Oh yeah, that makes sense.